_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,332 @@
+User-defined operators are a step towards that goal.
+
+#### Which Operators Can Be User-Defined?
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have unary `+`.
+
+#### Which Types Can Have User-Defined Operators?
+Currently, operators can only be defined for user-defined value types.
+We are also considering enabling them for structs and enums in future releases.
+There are no plans, for now, to allow defining operators for built-in value types, especially in cases where
+an operator is already provided by the compiler.
+
+Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
+is in line with how it already worked when used to attach individual functions.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
+#### What Are the Restrictions on User-Defined Operator Definitions?
+The operator definition must be a pure function, defined at the file level or inside a library.
+The type specified in `using for` must be the type of all parameters and the return value.
+The only exception are comparison operators, where the return type must be `bool`.
+
+A few points of note:
+- Being a `pure` function, the operator definition is allowed to make `pure` external calls.
+- Library functions are allowed but contract and interface functions are not.
+    This means that the definition cannot be virtual.
+- Private library functions are allowed but visible only within the library.
+- Operator definition may be a public library function.
+    External functions will eventually be allowed too
+    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
+- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
+    for selecting a specific overload, it is currently not possible to use such functions.
+- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
+    parameter and return value types happen to match the type specified in `using for`.
+
+The rules for `using for global` with operators are the same as for attached functions:
+it is only allowed at the file level, and the type must be defined in the current source unit and the
+effects extend to wherever the type is visible, not just the current source unit.
+
+In the next example, library `L` is used to group functions assigned to arithmetic operators for type `Int`.
+We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `=`.
```

Comment:
```suggestion
The operator definition must be a pure function, defined at the file level or inside a library.
The type specified in `using for` must be the type of all parameters and the return value.
The only exception are comparison operators, where the return type must be `bool`.

A few points of note:
- Being a `pure` function, the operator definition is allowed to make `pure` external calls.
- Library functions are allowed but contract and interface functions are not.
    This means that the definition cannot be virtual.
- Private library functions are allowed but visible only within the library.
- Operator definition may be a public library function.
    External functions will eventually be allowed too
    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
    for selecting a specific overload, it is currently not possible to use such functions.
- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their
    parameter and return value types happen to match the type specified in `using for`.

The rules for `using for global` with operators are the same as for the attached functions:
it is only allowed at the file level, and the type must be defined in the current source unit and the
effects extend to wherever the type is visible, not just the current source unit.

In the next example, library `L` is used to group functions assigned to arithmetic operators for type `Int`.
We also add more operators, with functions `unsub` and `equals` being respectively assigned to unary `-` and comparison `=`.
```

Reply:
This correction only applies if you're mentioning the functions here, if it's a generic thing, disregard it.

Reply:
Yeah, I meant "attached functions" as a general concept, not some specific functions.